src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4400,350 +4405,727 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
-                            if (containerOnly && !isContainer)
-                                continue;
+                            _ = sb.Append('\'');
+                            quotesAreNeeded = true;
+                        }
+                        else if (!useSingleQuoteEscapeRules && path[i].IsDoubleQuote())
+                        {
+                            _ = sb.Append('`');
+                            quotesAreNeeded = true;
+                        }
+                        break;
+                }
 
-                            if (!containerOnly && !isContainer && !CheckFileExtension(providerPath, extension))
-                                continue;
+                _ = sb.Append(path[i]);
+            }
 
-                            string tooltip = providerPath, listItemText = Path.GetFileName(providerPath);
-                            results.Add(new CompletionResult(completionText, listItemText,
-                                                             isContainer ? CompletionResultType.ProviderContainer : CompletionResultType.ProviderItem,
-                                                             tooltip));
+            return sb.ToString();
+        }
+
+        private static string EscapePath(string path, StringConstantType stringType, bool literalPath, out bool quotesAreNeeded)
+        {
+            var sb = new StringBuilder(path.Length);
+            bool useSingleQuoteEscapeRules = stringType is StringConstantType.SingleQuoted or StringConstantType.BareWord;
+            quotesAreNeeded = false;
+
+            for (int i = 0; i < path.Length; i++)
+            {
+                switch (path[i])
+                {
+                    case '#':
+                    case '&':
+                    case '-':
+                    case '@':
+                        if (i == 0 && stringType == StringConstantType.BareWord)
+                        {
+                            quotesAreNeeded = true;
                         }
-                        else
+                        break;
+
+                    case ' ':
+                    case ',':
+                    case ';':
+                    case '(':
+                    case ')':
+                    case '{':
+                    case '}':
+                    case '|':
+                        if (stringType == StringConstantType.BareWord)
                         {
-                            powerShellExecutionHelper
-                                .AddCommandWithPreferenceSetting("Microsoft.PowerShell.Management\\Get-Item")
-                                .AddParameter("LiteralPath", path);
-                            var items = powerShellExecutionHelper.ExecuteCurrentPowerShell(out exceptionThrown);
-                            if (items != null && items.Count == 1)
+                            quotesAreNeeded = true;
+                        }
+                        break;
+
+                    case '0':
```

Comment:
Since these blocks are already repeated from above, I wonder if it's better to have some helper functions to check as it'll be more descriptive and used in both methods.